File: src/visual-graph/builder/branch-container-key.ts
import type { SerializedScope } from "../../ir/model.js";
import { AST_TYPE } from "../../parser/ast-type.js";

export function branchContainerKey(scope: SerializedScope): string | null {
  const ctx = scope.blockContext;
  if (!ctx) {
    return null;
  }
  if (ctx.parentType === AST_TYPE.SwitchStatement && ctx.key === "cases") {
    return `switch:${scope.upper ?? ""}:${ctx.parentSpanOffset}`;
  }
  if (
    ctx.parentType === AST_TYPE.IfStatement &&
    (ctx.key === "consequent" || ctx.key === "alternate")
  ) {
    const root =
      ctx.kind === "other" && ctx.ifChainRootOffset !== undefined
        ? ctx.ifChainRootOffset
        : ctx.parentSpanOffset;
    return `if:${scope.upper ?? ""}:${root}`;
  }
  if (
    ctx.parentType === AST_TYPE.TryStatement &&
    (ctx.key === "block" || ctx.key === "handler")
  ) {
    return `try:${scope.upper ?? ""}:${ctx.parentSpanOffset}`;
  }
  return null;
}

File: crates/unsnarl-visual-graph/src/builder/branch_container_key.rs
//! Computes a string key that groups every branch of a single
//! `if` / `switch` / `try` statement under a stable hashable
//! identifier. The key blends the parent scope id with the
//! statement's `parentSpanOffset` (or, for `else if` chains, the
//! `ifChainRootOffset`) so branches that should merge into one
//! container share the same key.

use unsnarl_ir::scope::block_context::BlockContext;
use unsnarl_ir::scope::block_context_kind::BlockContextKind;
use unsnarl_ir::serialized::SerializedScope;
use unsnarl_oxc_parity::AstType;

pub fn branch_container_key(scope: &SerializedScope) -> Option<String> {
    let ctx = scope.block_context.as_ref()?;
    let upper = scope.upper.as_ref().map(|u| u.value()).unwrap_or("");

    match ctx {
        BlockContext::CaseClause(c) => {
            if matches!(c.parent_type(), AstType::SwitchStatement) && c.key() == "cases" {
                return Some(format!("switch:{upper}:{}", c.parent_span_offset().0));
            }
            None
        }
        BlockContext::Other(c) => {
            if matches!(c.parent_type(), AstType::IfStatement)
                && (c.key() == "consequent" || c.key() == "alternate")
            {
                let root = if matches!(c.kind(), BlockContextKind::Other) {
                    c.if_chain_root_offset()
                        .map(|o| o.0)
                        .unwrap_or_else(|| c.parent_span_offset().0)
                } else {
                    c.parent_span_offset().0
                };
                return Some(format!("if:{upper}:{root}"));
            }
            if matches!(c.parent_type(), AstType::TryStatement)
                && (c.key() == "block" || c.key() == "handler")
            {
                return Some(format!("try:{upper}:{}", c.parent_span_offset().0));
            }
            None
        }
    }
}

#[cfg(test)]
#[path = "branch_container_key_test.rs"]
mod branch_container_key_test;
